metricproviders/datadog/datadog.go

```diff
@@ -0,0 +1,172 @@
+	return measurement
+}
+
+// Terminate should not be used with the datadog provider since all the work should occur in the Run method
+func (p *Provider) Terminate(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Datadog provider should not execute the Terminate method")
+	return measurement
+}
+
+// GarbageCollect is a no-op for the datadog provider
+func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, limit int) error {
+	return nil
+}
+
+func (p *Provider) processResponse(metric v1alpha1.Metric, response dd.Series) (string, v1alpha1.AnalysisPhase, error) {
+	length := len(response.Points)
+	switch {
+	case length < 1:
+		return "", v1alpha1.AnalysisPhaseInconclusive, nil
+	// If we get 1 or more points, select the first point
+	case length >= 1:
+		result := *response.Points[0][1]
+		valueStr := fmt.Sprintf("%f", result)
+		newStatus := evaluate.EvaluateResult(result, metric, p.logCtx)
+		return valueStr, newStatus, nil
+	// TODO (cabrinha) add other response types
+	default:
+		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("Datadog metric type not supported")
+	}
+}
+
+// NewDatadogProvider creates a new Datadog client
+func NewDatadogProvider(client dd.Client, logCtx log.Entry) *Provider {
+	return &Provider{
+		logCtx: logCtx,
+		client: client,
+	}
+}
+
+// NewDatadogAPI generates a datadog API from the metric configuration
+func NewDatadogAPI(metric v1alpha1.Metric, kubeclientset kubernetes.Interface) (*dd.Client, error) {
+	ns := Namespace()
+	secret, err := kubeclientset.CoreV1().Secrets(ns).Get(DatadogSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(secret.Data[`datadog-api-key`]) > 0 && len(secret.Data[`datadog-app-key`]) > 0 {
+		apiKey := fmt.Sprintf("%s", secret.Data[`datadog-api-key`])
+		appKey := fmt.Sprintf("%s", secret.Data[`datadog-app-key`])
+		client := dd.NewClient(apiKey, appKey)
+
+		if metric.Provider.Datadog.BaseURL != "" {
+			client.SetBaseUrl(metric.Provider.Datadog.BaseURL)
+		}
+
+		_, err := client.Validate()
+		if err != nil {
+			return nil, err
+		}
+		return client, nil
+	}
+	return nil, errors.New("failed to make client: no datadog API or App keys found")
+}
+
+// Namespace will return $POD_NAMESPACE if set, otherwise "argo-rollouts"
+func Namespace() string {
+	// This way assumes you've set the POD_NAMESPACE environment variable using the downward API.
+	// This check has to be done first for backwards compatibility with the way InClusterConfig was originally set up
+	if ns, ok := os.LookupEnv("POD_NAMESPACE"); ok {
+		return ns
+	}
+	// Fall back to the namespace associated with the service account token, if available
+	if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
+		if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
+			return ns
+		}
+	}
+	return "argo-rollouts"
+}
```

Comment:
Can you move this method to the utils/controller package and have the Wavefront and Datadog packages call this method instead?